diff --git a/lib/Target/AMDGPU/AMDGPUAsmPrinter.cpp b/lib/Target/AMDGPU/AMDGPUAsmPrinter.cpp
index d704a0fae0d..b988d540000 100644
--- a/lib/Target/AMDGPU/AMDGPUAsmPrinter.cpp
+++ b/lib/Target/AMDGPU/AMDGPUAsmPrinter.cpp
@@ -489,7 +489,13 @@ bool AMDGPUAsmPrinter::runOnMachineFunction(MachineFunction &MF) {
       OutStreamer->EmitBytes(StringRef(Comment));
     }
   }
-
+  const GCNSubtarget &STI = MF.getSubtarget<GCNSubtarget>();
+  auto OccVGPRs = STI.getOccupancyWithNumVGPRs(CurrentProgramInfo.NumVGPR);
+  auto OccSGPRs = STI.getOccupancyWithNumSGPRs(CurrentProgramInfo.NumSGPR);
+  auto OccLDS = STI.getOccupancyWithLocalMemSize(MF);
+ 
+  auto Occ = std::min(OccVGPRs, std::min(OccSGPRs, OccLDS));
+  dbgs() << "Final occupancy for function " << MF.getName() << ":" << Occ << "\n";
   return false;
 }
 
